ユーザー管理機能

miracletea opened the pull request:
# What
ユーザー管理機能、およびテストコードを作成しました。
# Why
ユーザー管理機能の実装のため。

【Gyazo】
・必要な情報を適切に入力して「会員登録」ボタンを押すと、ユーザーの新規登録ができる動画
https://gyazo.com/3d04b302eebb60f954e7a2a4fc6efc87
・入力に問題のある状態で「会員登録」ボタンが押された場合、ユーザーの新規登録はできず、エラーメッセージが出力される動画
https://gyazo.com/796dc98e76fe1e5f588ad2f153ab8b01
・必要な情報を適切に入力して「ログイン」ボタンを押すと、ユーザーのログインができる動画
https://gyazo.com/5191b5e2d4f57ad0f28e63d58f5c995e
・入力に問題のある状態で「ログイン」ボタンが押された場合、ユーザーのログインはできず、エラーメッセージが出力される動画
https://gyazo.com/3169a2c90a576daa6a2d8213847a9ee2
・ログアウトに成功する動画
https://gyazo.com/9575b492011c606eed9fc9c1f68cc27c
・テストに成功する画像
https://gyazo.com/8fbc2839f5dcc0c597efa84a736aa838

miracletea committed:
ユーザー7新規登録、ログイン、ログアウトの実装

miracletea committed:
ユーザ管理機能のテストコード実装

CHURO-MAU commented on app/controllers/items_controller.rb:
```diff
@@ -1,6 +1,14 @@
 class ItemsController < ApplicationController
+  before_action :move_to_index, except:[:index, :show]
```
２行目
一旦コメントアウトにしておきましょう。

理由：
ユーザー管理機能の実装要件ではないためになります。こちらの実装が必要となる機能実装時にコメントアウトを外してコミット・プッシュしていきましょう。

CHURO-MAU commented on app/controllers/items_controller.rb:
```diff
@@ -1,6 +1,14 @@
+    unless user_signed_in?
+      redirect_to action: :index
+    end
+  end
```
９〜１３行目
こちらも同様に一旦コメントアウトにしておきましょう。

CHURO-MAU commented on app/models/user.rb:
```diff
@@ -0,0 +1,14 @@
+
+  validates :nickname, presence: true
+  validates :last_name, presence: true, format: { with: /\A[ぁ-んァ-ン一-龥]/ }
+  validates :first_name, presence: true, format: { with: /\A[ぁ-んァ-ン一-龥]/ }
```
８，９行目
正規表現に誤りがございますので、修正しましょう。

現状ですと、山田Aのように名前に英字が含まれていても保存されてしまうと思われます。また、佐々木やトニーのように々やー（長音記号）が許容できておりません。

CHURO-MAU commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,108 @@
+
+<%# 「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
+<%= form_with model: @user, url: user_registration_path, class: 'registration-main', local: true do |f| %>
+<%# //「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
```
３，５行目
「〜しましょう」で終わる実装案内コメントは実装完了後に削除しましょう。
企業提出時に実装案内コメントが残っていないようにするためです。

CHURO-MAU commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,108 @@
+
+  <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
+  <%= render 'shared/error_messages', model: f.object %>
+  <%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
１３，１５行目
こちらも同様に削除しましょう。

CHURO-MAU commented on app/views/shared/_header.html.erb:
```diff
@@ -16,12 +16,13 @@
+        <li><%= link_to 'ログイン', user_session_path, class: "login" %></li>
+        <li><%= link_to '新規登録', new_user_registration_path, class: "sign-up" %></li>
+      <% end %>
       <%# //deviseを導入できたら、ログインの有無で表示が変わるように分岐しましょう%>
```
１８，２６行目
こちらも同様に削除しましょう。

CHURO-MAU commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,15 @@
+
+    nickname              {Faker::Name.initials(number: 2)}
+    email                 {Faker::Internet.email}
+    password              {Faker::Internet.password(min_length: 6)}
```
passwordカラムを修正しましょう。

現状、必ずしも半角英数混合になるかどうかが不明となる記述です。
(Fakerのみの場合、英字のみのパスワードが生成されてしまうことがあるためです。)
確実に半角英数混合の値を生成するため、『+』の記号を使用し、文字列を組み合わせることで期待する値を生成する内容に修正していきましょう。

CHURO-MAU commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,85 @@
+    @user = FactoryBot.build(:user)
+  end
+
+  describe 'ユーザー新規登録' do
```
正常系テストの実装を行いましょう。

■理由
モデルの単体テスコードの実装要件に含まれているためです。
※テストコードはcontextを用いて正常系と異常系の二つにブロック分けの上、実装を行いましょう。

CHURO-MAU commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,85 @@
+    it 'passwordが空では登録できない' do
+      @user.password = ''
+      @user.valid?
+      expect(@user.errors.full_messages).to include("Password can't be blank", "Password confirmation doesn't match Password", "Password is invalid")
```
２３行目
include()内には、このテストで期待する適切なメッセージを１つ記述しましょう。（現状３つのメッセージが記述されています）

**※その他のテストでもinclude()内に複数のメッセージが記述されている箇所がございます。併せて修正しておきましょう。**

CHURO-MAU commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,85 @@
+  end
+
+  describe 'ユーザー新規登録' do
+    #ユーザー新規登録についてのテストコードを記述します。
```
以下の異常系テストを追加しましょう。
■パスワード
・英字のみのパスワードでは登録できない
・数字のみのパスワードでは登録できない
・全角文字を含むパスワードでは登録できない
↑パスワードのテストは半角英数字混合での入力でなければ登録できないを確かめるために、上記のように細分化してテストを行いましょう。

■姓・名（全角）
・姓（全角）に半角文字が含まれていると登録できない
・名（全角）に半角文字が含まれていると登録できない

■姓・名（カナ）
・姓（カナ）にカタカナ以外の文字（平仮名・漢字・英数字・記号）が含まれていると登録できない
・名（カナ）にカタカナ以外の文字（平仮名・漢字・英数字・記号）が含まれていると登録できない

次回レビュー依頼時には、「最新のテスト結果の画像」もご提出をお願いいたします。
「最新のテスト結果の画像」は、プルリクエストに添付いただければと思います。

CHURO-MAU reviewed (commented):
レビュー依頼ありがとうございます！

コメントいたしましたので、ご確認お願いいたします。
修正後は、再度レビュー依頼フォームのご提出をお願いいたします。

miracletea committed:
itemsコントローラの2行目コメントアウト

app/controllers/items_controller.rb
@@ -1,6 +1,14 @@
class ItemsController < ApplicationController
  before_action :move_to_index, except:[:index, :show]
@CHURO-MAU CHURO-MAU 11 hours ago
２行目
一旦コメントアウトにしておきましょう。

理由：
ユーザー管理機能の実装要件ではないためになります。こちらの実装が必要となる機能実装時にコメントアウトを外してコミット・プッシュしていきましょう。

miracletea committed:
itemsコントローラの9-13行目コメントアウト

app/controllers/items_controller.rb
Comment on lines +9 to +13
  def move_to_index
    unless user_signed_in?
      redirect_to action: :index
    end
  end

miracletea committed:
last_name、first_nameの正規表現の見直し

app/models/user.rb
Comment on lines +8 to +9
  validates :last_name, presence: true, format: { with: /\A[ぁ-んァ-ン一-龥]/ }
  validates :first_name, presence: true, format: { with: /\A[ぁ-んァ-ン一-龥]/ }
@CHURO-MAU CHURO-MAU 11 hours ago
８，９行目
正規表現に誤りがございますので、修正しましょう。

現状ですと、山田Aのように名前に英字が含まれていても保存されてしまうと思われます。また、佐々木やトニーのように々やー（長音記号）が許容できておりません。

miracletea committed:
記述コメントの削除

app/views/devise/registrations/new.html.erb
Comment on lines +3 to +5
<%# 「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
<%= form_with model: @user, url: user_registration_path, class: 'registration-main', local: true do |f| %>
<%# //「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
@CHURO-MAU CHURO-MAU 11 hours ago
３，５行目
「〜しましょう」で終わる実装案内コメントは実装完了後に削除しましょう。
企業提出時に実装案内コメントが残っていないようにするためです。

miracletea committed:
記述コメントの削除

app/views/devise/registrations/new.html.erb
Comment on lines +13 to +15
  <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
  <%= render 'shared/error_messages', model: f.object %>
  <%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
@CHURO-MAU CHURO-MAU 11 hours ago
１３，１５行目
こちらも同様に削除しましょう。

miracletea committed:
記述コメントの削除

app/views/shared/_header.html.erb
Comment on lines 18 to 26
      <%# deviseを導入できたら、ログインの有無で表示が変わるように分岐しましょう%>
      <%# <li><%= link_to current_user.nickname, "#", class: "user-nickname" </li> %>
      <%# <li><%= link_to 'ログアウト', "#", data: {turbo_method: :delete}, class: "logout" </li> %>

      <li><%= link_to 'ログイン', "#", class: "login" %></li>
      <li><%= link_to '新規登録', "#", class: "sign-up" %></li>
      <% if user_signed_in? %>
        <li><%= link_to current_user.nickname, "#", class: "user-nickname" %></li>
        <li><%= link_to 'ログアウト', destroy_user_session_path, data: {turbo_method: :delete}, class: "logout" %></li>
      <% else %>
        <li><%= link_to 'ログイン', user_session_path, class: "login" %></li>
        <li><%= link_to '新規登録', new_user_registration_path, class: "sign-up" %></li>
      <% end %>
      <%# //deviseを導入できたら、ログインの有無で表示が変わるように分岐しましょう%>
@CHURO-MAU CHURO-MAU 11 hours ago
１８，２６行目
こちらも同様に削除しましょう。

miracletea committed:
Fakerのpasswordカラムの修正

spec/factories/users.rb

    nickname              {Faker::Name.initials(number: 2)}
    email                 {Faker::Internet.email}
    password              {Faker::Internet.password(min_length: 6)}
@CHURO-MAU CHURO-MAU 11 hours ago
passwordカラムを修正しましょう。

現状、必ずしも半角英数混合になるかどうかが不明となる記述です。
(Fakerのみの場合、英字のみのパスワードが生成されてしまうことがあるためです。)
確実に半角英数混合の値を生成するため、『+』の記号を使用し、文字列を組み合わせることで期待する値を生成する内容に修正していきましょう。

miracletea committed:
正常系テストの実装

spec/models/user_spec.rb
    @user = FactoryBot.build(:user)
  end

  describe 'ユーザー新規登録' do
@CHURO-MAU CHURO-MAU 12 hours ago
正常系テストの実装を行いましょう。

■理由
モデルの単体テスコードの実装要件に含まれているためです。
※テストコードはcontextを用いて正常系と異常系の二つにブロック分けの上、実装を行いましょう。

miracletea committed:
include()内の不要なメッセージを削除

spec/models/user_spec.rb
    it 'passwordが空では登録できない' do
      @user.password = ''
      @user.valid?
      expect(@user.errors.full_messages).to include("Password can't be blank", "Password confirmation doesn't match Password", "Password is invalid")
@CHURO-MAU CHURO-MAU 12 hours ago
２３行目
include()内には、このテストで期待する適切なメッセージを１つ記述しましょう。（現状３つのメッセージが記述されています）

※その他のテストでもinclude()内に複数のメッセージが記述されている箇所がございます。併せて修正しておきましょう。

miracletea committed:
異常系テストの追加

以下の異常系テストを追加しましょう。
■パスワード
・英字のみのパスワードでは登録できない
・数字のみのパスワードでは登録できない
・全角文字を含むパスワードでは登録できない
↑パスワードのテストは半角英数字混合での入力でなければ登録できないを確かめるために、上記のように細分化してテストを行いましょう。

■姓・名（全角）
・姓（全角）に半角文字が含まれていると登録できない
・名（全角）に半角文字が含まれていると登録できない

■姓・名（カナ）
・姓（カナ）にカタカナ以外の文字（平仮名・漢字・英数字・記号）が含まれていると登録できない
・名（カナ）にカタカナ以外の文字（平仮名・漢字・英数字・記号）が含まれていると登録できない

miracletea commented:
ご確認ありがとうございます、コメント反映いたしました。

【Gyazo】
・テストに成功する画像
https://gyazo.com/004f61a1a815d20287ba9b49d1086c8d

minachanchan commented on app/views/devise/sessions/new.html.erb:
```diff
@@ -0,0 +1,34 @@
+<%= render "shared/second-header"%>
+
+<%= form_with model: @user, url: new_user_session_path, class: 'registration-main', local: true do |f| %>
```
```erb
<%= form_with model: @user, url: new_user_session_path, class: 'registration-main', local: true do |f| %>
```
`new_user_session_path`はsessions#newへのパスでございます。
createへのパスに修正しましょう！

minachanchan commented on app/views/shared/_header.html.erb:
```diff
@@ -15,14 +15,13 @@
+        <li><%= link_to current_user.nickname, "#", class: "user-nickname" %></li>
+        <li><%= link_to 'ログアウト', destroy_user_session_path, data: {turbo_method: :delete}, class: "logout" %></li>
+      <% else %>
+        <li><%= link_to 'ログイン', user_session_path, class: "login" %></li>
```
```erb
        <li><%= link_to 'ログイン', user_session_path, class: "login" %></li>
Comment
```
こちらはsessions#createでございます。
newアクションへのパスに修正しましょう！

minachanchan reviewed (commented):
実装お疲れ様です！レビューのご依頼ありがとうございます。
着実に完成に近づいておりますね！

コメントいたしましたので、ご確認頂けますと幸いでございます。

実装完了まであと少し、一緒に頑張りましょう！
修正の後、再度レビューのご依頼頂けますようお願い申し上げます。

miracletea committed:
sessionパスの見直し

app/views/devise/sessions/new.html.erb
@@ -0,0 +1,34 @@
<%= render "shared/second-header"%>
<%= form_with model: @user, url: new_user_session_path, class: 'registration-main', local: true do |f| %>
@minachanchan minachanchan 2 hours ago
<%= form_with model: @user, url: new_user_session_path, class: 'registration-main', local: true do |f| %>
new_user_session_pathはsessions#newへのパスでございます。
createへのパスに修正しましょう！

app/views/shared/_header.html.erb
        <li><%= link_to current_user.nickname, "#", class: "user-nickname" %></li>
        <li><%= link_to 'ログアウト', destroy_user_session_path, data: {turbo_method: :delete}, class: "logout" %></li>
      <% else %>
        <li><%= link_to 'ログイン', user_session_path, class: "login" %></li>
@minachanchan minachanchan 2 hours ago
        <li><%= link_to 'ログイン', user_session_path, class: "login" %></li>
Comment
こちらはsessions#createでございます。
newアクションへのパスに修正しましょう！

miracletea commented:
ご確認ありがとうございます、コメント反映いたしました。

minachanchan reviewed (commented):
レビューのご依頼ありがとうございます。

実装内容を確認させていただきました。
**LGTMとさせていただきます。実装お疲れ様でした！**

このブランチをマージし、masterブランチ（もしくはmainブランチ）でプルしましょう。
内容が反映されていることを確認し、次の実装・学習に入りましょう！
引き続き応援しております！
